File: Year-2022/Day02/day02.py
from myutils import run_test
from data_structures import CircularDLList


def map_move(move: str):
    if move == "A" or move == "X":
        return "R"
    elif move == "B" or move == "Y":
        return "P"
    elif move == "C" or move == "Z":
        return "S"
    else:
        raise ValueError("Invalid move")


class RockPaperScissors:

    moves = CircularDLList("RPS")
    move_value = {"R": 1, "P": 2, "S": 3}

    @classmethod
    def round_outcome(cls, move_1, move_2):
        if move_1 == move_2:
            return 3
        elif move_2 == cls.moves[move_1].next.data:
            return 6
        else:
            return 0

    @classmethod
    def round_score(cls, move_1, move_2):
        return cls.move_value[move_2] + cls.round_outcome(move_1, move_2)

    @classmethod
    def fix_round(cls, move_1, outcome):
        score = 0
        if outcome == "X":  # loss
            my_move = cls.moves[move_1].previous.data
        elif outcome == "Y":  # draw
            score += 3
            my_move = move_1
        else:  # win
            score += 6
            my_move = cls.moves[move_1].next.data
        score += cls.move_value[my_move]
        return score


def read_input(text_file):
    rounds = []
    with open(text_file) as f:
        for line in f.readlines():
            rounds.append(line.split())
    return rounds


@run_test("day02_test.txt", 15)
def day02_part1(day_input: str):
    rounds = read_input(day_input)
    score = 0
    for r in rounds:
        score += RockPaperScissors.round_score(map_move(r[0]), map_move(r[1]))
    return score


@run_test("day02_test.txt", 12)
def day02_part2(day_input: str):
    rounds = read_input(day_input)
    score = 0
    for r in rounds:
        score += RockPaperScissors.fix_round(map_move(r[0]), r[1])
    return score


if __name__ == "__main__":
    print("Part 1:\t", day02_part1("day02.txt"))
    print("Part 2:\t", day02_part2("day02.txt"))

File: Year-2022/perl/Day02/day02.pl
use strict;
use warnings;
use autodie;

use feature 'say';

# A X - Rock 1
# B Y - Paper 2
# C Z - Scissors 3
# Win 3

my $file;

open($file, '<', 'Day02/day02.txt') or die "Couldn't open file!";
my %score_hash = ('X' => 1, 'Y' => 2, 'Z' => 3, 'A' => 1, 'B' => 2, 'C' => 3);
my %win_hash = ('X' => 'C', 'Y' => 'A', 'Z' => 'B');
my $score = 0;
my $opp_move;
my $my_move;
while (my $line = <$file>) {
    chomp $line;
    ($opp_move, $my_move) = split(' ', $line);
    $score += $score_hash{$my_move};
    if ($score_hash{$my_move} eq $score_hash{$opp_move}) {
        $score += 3;
    } elsif ($opp_move eq $win_hash{$my_move}) {
        $score += 6;
    }
}
say $score;
close($file);

open($file, '<', 'Day02/day02.txt') or die "Couldn't open file!";
my %cheat_hash = ('X' => 0, 'Y' => 3, 'Z' => 6);
$score = 0;
while (my $line = <$file>) {
    chomp $line;
    ($opp_move, $my_move) = split(' ', $line);
    $score += $cheat_hash{$my_move};
    if ($my_move eq 'X') {
        $score += 1 + (($score_hash{$opp_move} - 2) % 3)
    } elsif ($my_move eq 'Y') {
        $score += $score_hash{$opp_move};
    } else {
        $score += 1 + $score_hash{$opp_move} % 3;
    }
}
say $score;
close($file);
